UnityProject/Assets/Scripts/Weapons/Projectiles/MovingProjectile.cs

```diff
@@ -0,0 +1,80 @@
+﻿using Container.Gun;
+using UnityEngine;
+
+namespace Weapons.Projectiles
+{
+	/// <summary>
+	/// This script is used for actually moving projectiles sprite and raycasting
+	/// Reason why it was done in local space and on a separate game object child it this
+	/// LocalTrailRanderer cannot properly draw line in World space if it's done on a moving matrix
+	/// </summary>
+	public class MovingProjectile : MonoBehaviour
+	{
+		private Bullet projectile;
+		private LayerMaskData maskData;
+		private Transform thisTransform;
+
+		private Vector3 mPrevPos;
+
+		private float velocity;
+
+		private void Awake()
+		{
+			projectile = GetComponentInParent<Bullet>();
+			maskData = projectile.MaskData;
+			thisTransform = transform;
+		}
+
+		/// <summary>
+		/// Method to rotate and reset projectile position in local space
+		/// </summary>
+		/// <param name="direction"> Direction to travel </param>
+		/// <param name="velocity"> Projectile speed </param>
+		public void SetUpBulletTransform(Vector2 direction, float velocity)
+		{
+			this.velocity = velocity;
+
+			var startPosition =  Vector3.zero;
+			var startRotation = Quaternion.AngleAxis(-Mathf.Atan2(direction.x, direction.y) * Mathf.Rad2Deg, Vector3.forward);
+
+			thisTransform.localPosition = startPosition;
+			thisTransform.rotation = startRotation;
+		}
+
+		private void Update()
+		{
+			CachePreviousPosition();
+
+			MoveBullet();
+
+			var hit = Raycast();
+
+			projectile.ProcessRaycastHit(hit);
```

Comment:
why not inline `hit` variable?

Reply:
>I see a lot of GetComponent<> calls 

I think it is fine. Unity is component system based and you will have to do at least one `GetComponent` to apply game logic. But yeah, there are ways to make it better. For tile interactions you can just get `MetaTileMap` and pass it to all behaviours which will need it. For Damaging interactions, it would be much easier if we had interface for damaging objects in game instead what we have now. You would just need to grab `IApplyDamage` for example from collider and do the dark business. 

>Also curious if it's possible to make certain projectiles emit light

Yes, and yes they do that.

Reply:
You mean : `projectile.ProcessRaycastHit( Raycast());` ?
I think i did it cuz it was easy to see in debug what hit is, but yeah.

Reply:
You want to inline it?